### PR TITLE
Feature/pov kotlin rewrite core lib logic

### PR DESCRIPTION
**Pull request by ItsBruno**
## Summary

This PR continues on [PR](https://github.com/infinum/Android-Prince-of-Versions/pull/118) and refactors the remaining core library logic.

## Changes

- Refactored core library classes to Kotlin (Lib to app interface class, presenter, interactor, JSON parser, Executer, Config, etc.)
- After finishing the rewrite and having all tests passed, deleted Java tests and renamed Kotlin refactored ones.
- Fixed bug in UpdaterCall caused by missing return statements in isCanceled if blocks in enqueue methods.
- Uncommented tests that test the case mentioned in the previous point. 
- Reimplemented CheckResult as DataClass with `@ConsistentCopyVisibility` annotation.
- Deleted ApplicationVersionProvider in favour for a simple lambda function.

### Type

- [ ] **Feature**: This pull request introduces a new feature.
- [ ] **Bug fix**: This pull request fixes a bug.
- [X] **Refactor**: This pull request refactors existing code.
- [ ] **Documentation**: This pull request updates documentation.
- [ ] **Other**: This pull request makes other changes.

#### Additional information

- [ ] This pull request introduces a **breaking change**.

## Checklist

- [X] I have performed a self-review of my own code.
- [X] I have tested my changes, including edge cases.
- [ ] I have added necessary tests for the changes introduced (if applicable).
- [ ] I have updated the documentation to reflect my changes (if applicable).

## Additional notes

Question for reviewers - Will the rewrite to KMP sunset this repository - in other words, do we need to update the README and other documentation with this rewrite?

### CODE REVIEW

File: prince-of-versions/src/main/java/co/infinum/princeofversions/PrinceOfVersionsCallbackExecutor.kt [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import android.os.Handler
+import android.os.Looper
+import java.util.concurrent.Executor
+
+/**
+ * An implementation of [Executor] that runs tasks on the Android main thread.
+ */
+class PrinceOfVersionsCallbackExecutor : Executor {
+
+    private val handler = Handler(Looper.getMainLooper())
```
[the remark this answers is not in the document]
Comment (KCeh) [reply prince-of-versions/src/main/java/co/infinum/princeofversions/PrinceOfVersionsCallbackExecutor.kt:12]: Nice catch 👍

**Review by KCeh — COMMENTED** [review]
Looks good, let's just improve Sonar warnings
For critical sonar errors you can define const val
For major ones, should be straight forward

As for improvements in interfaces and patters keep this in mind when creating KMP code:
We should try to keep public API the same (unless there is a good reason why it can't be).
But all internal APIs that can be more in spirit of Kotlin (suspend functions, non builder patterns for creating data class instance etc.) are welcomed improvements.

> Question for reviewers - Will the rewrite to KMP sunset this repository - in other words, do we need to update the README and other documentation with this rewrite?

Probably yes, but we won't archive this repo immediately.
👍 for thinking ahead.
Librarian (me) will take care of those things when time comes

File: prince-of-versions/src/main/java/co/infinum/princeofversions/PrinceOfVersionsCallbackExecutor.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (KCeh) [reply prince-of-versions/src/main/java/co/infinum/princeofversions/PrinceOfVersionsCallbackExecutor.kt:12]: hmmm @antunflas can you jump in with additional insight?

**Review by KCeh — APPROVED** [review]

File: prince-of-versions/src/main/java/co/infinum/princeofversions/JsonConfigurationParser.kt [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+     */
+    @Suppress("kotlin:S6531")
+    private fun saveUpdateValues(update: JSONObject, builder: PrinceOfVersionsConfig.Builder) {
+        if (!update.isNull(MINIMUM_VERSION)) {
+            val value = update[MINIMUM_VERSION]
+            if (value is Int) {
+                builder.withMandatoryVersion(value)
+            } else {
+                throw IllegalArgumentException("In update configuration $MINIMUM_VERSION it should be int, but the actual value is $value")
+            }
```
[the remark this answers is not in the document]
Comment (KCeh) [reply prince-of-versions/src/main/java/co/infinum/princeofversions/JsonConfigurationParser.kt:110]: Kudos on good comments
